web3/utils/abi.py

```diff
@@ -158,46 +164,153 @@ def _get_fallback_function_abi(contract_abi: ABI) -> ABIFallback:
     arguments or no functions are identified.
     """
-    diagnosis = "\n"
-    if arg_count_matches == 0:
-        diagnosis += "Function invocation failed due to improper number of arguments."
-    elif encoding_matches == 0:
-        diagnosis += "Function invocation failed due to no matching argument types."
-    elif encoding_matches > 1:
-        diagnosis += (
-            "Ambiguous argument encoding. "
-            "Provided arguments can be encoded to multiple functions "
-            "matching this call."
-        )
+    name = get_name_from_abi_element_identifier(abi_element_identifier)
+    abis_matching_names = filter_abi_by_name(name, abi)
+    abi_signatures_matching_names = [
+        abi_to_signature(abi) for abi in abis_matching_names
+    ]
+    abis_matching_arg_count = [
+        abi_to_signature(abi)
+        for abi in _filter_by_argument_count(num_args, abis_matching_names)
+    ]
+    num_abis_matching_arg_count = len(abis_matching_arg_count)
 
-    collapsed_args = _extract_argument_types(*args)
-    collapsed_kwargs = dict(
-        {(k, _extract_argument_types([v])) for k, v in kwargs.items()}
-    )
+    if abi_codec is None:
+        abi_codec = ABICodec(default_registry)
 
-    return (
-        f"\nCould not identify the intended function with name "
-        f"`{abi_element_identifier}`, positional arguments with type(s) "
-        f"`({collapsed_args})` and keyword arguments with type(s) "
-        f"`{collapsed_kwargs}`."
-        f"\nFound {len(matching_function_signatures)} function(s) with the name "
-        f"`{abi_element_identifier}`: {matching_function_signatures}{diagnosis}"
+    error = "ABI Not Found!\n"
+    if num_matches == 0 and num_abis_matching_arg_count == 0:
+        error += (
+            f"No declaration found for `{str(abi_element_identifier)}` with "
+            f"{num_args} argument(s).\n"
+        )
+    elif num_matches > 1 or num_abis_matching_arg_count > 1:
+        error += f"Multiple elements were found that accept {num_args} argument(s).\n"
+    elif num_abis_matching_arg_count == 1:
+        error += (
+            f"Found {num_abis_matching_arg_count} element(s) named `{name}` that "
+            f"accept {num_args} argument(s).\n"
+            "The provided arguments do not match the expected types.\n"
+        )
+    elif num_matches == 0:
+        error += (
+            "Unable to find an element that matches the provided identifier and "
+            "argument types.\n"
+        )
+    arg_types = _extract_argument_types(*args)
+    kwarg_types = dict({(k, _extract_argument_types([v])) for k, v in kwargs.items()})
+    error += (
+        f"Provided argument types: ({arg_types})\n"
+        f"Provided keyword argument types: {kwarg_types}\n\n"
+        f"Tried to find a matching ABI element named `{name}`, but encountered "
+        "the following problems:\n"
     )
 
+    for abi_signature in abi_signatures_matching_names:
+        error += _build_abi_input_error(
```

Comment:
@reedsa I approbed but I think we need to tighten up the messaging a little bit. It looks like we only append problems when there are matching names. Perhaps we should make this code block a bit more like:

```py
if abi_signatures_matching_names:
    error += f"Tried to find a matching ABI element named `{name}`, but encountered the following problems:\n"

    for abi_signature in abi_signatures_matching_names:
        error += _build_abi_input_error(...)
```

Because if I run something like the following:

```bash
$ python -c "from web3.utils import get_abi_element;get_abi_element({}, 'nonexistent')"
```

I will get "encountered the following problems:" with no "problems" after it, just a blank.